draft-ietf-ppm-dap.md

```diff
@@ -1577,22 +1595,45 @@ Finally, the Helper creates an `AggregationJobResp` to send to the Leader. This
 message is structured as follows:
 
 ~~~
+enum {
+  processing(0),
+  finished(1),
+} AggregationJobStatus;
+
 struct {
-  PrepareResp prepare_resps<1..2^32-1>;
+  AggregationJobStatus status;
+  select (AggregationJobResp.status) {
+    case processing: Empty;
+    case finished:   PrepareResp prepare_resps<0..2^32-1>;
+  };
 } AggregationJobResp;
 ~~~
 
 where `prepare_resps` are the outbound prep steps computed in the previous step.
 The order MUST match `AggregationJobInitReq.prepare_inits`.
 
-The Helper responds to the Leader with HTTP status code 201 Created and a body
-consisting of the `AggregationJobResp`, with media type
+The Helper responds to the Leader with HTTP status 201 Created, a body
+consisting of the `AggregationJobResp`, and the media type
 "application/dap-aggregation-job-resp".
 
-Changing an aggregation job's parameters is illegal, so further requests to
-`PUT /tasks/{tasks}/aggregation_jobs/{aggregation-job-id}` for the same
-`aggregation-job-id` but with a different `AggregationJobInitReq` in the body
-MUST fail with an HTTP client error status code.
+Depending on the task parameters, processing an aggregation job may take some
+time, so the Helper MAY defer computation to a background process. It does so
+by responding with the field `status` set to `processing` and a Location header
+field set to the relative reference
+`/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0`. The Leader
+then polls the Helper by making HTTP GET requests to the aforementioned
+Location. The Helper responds to GET requests with HTTP status 200 and the
+`status` field reflecting the current state of the job. When the aggregation
+job is `processing`, the response SHOULD include a Retry-After header field to
+suggest a polling interval to the Leader.
```

Comment:
Consider spelling out the status code if a GET is made before the PUT. This might happen for example if the Leader crashes during the polling loop.

Reply:
Here and below (continuation).

Reply:
Is this scenario assuming that a Leader moves onto polling without check-pointing its progress? That is, it does
```rust
let mut aggregation_job_resp = post(aggregation_job);
while !aggregation_job_resp.status.is_done() {
    // n.b. no checkpointing to datastore is done
    aggregation_job_resp = get(aggregation_job_resp.polling_location);
}
```
So if it crashes during this function, it'll restart from the `post()` call?

I think this is already specified more or less by:
```
For further requests with
the same `AggregationJobInitReq` in the body, the Helper SHOULD respond as it
did for the original `AggregationJobInitReq`, or otherwise fail with an HTTP
client error status code.
```

Since it'll be repeating the AggreationJobInitReq.

(Let me know if I'm misunderstanding the scenario).

Reply:
I was thinking more of a misbehaving Leader trying to GET an aggregation job step before kicking it off with a PUT/GET. Thinking about it more, this is just a plain old 404. I don't think it's worth documenting.